### PR TITLE
Gas optimization on erc-20

### PR DESCRIPTION
**Pull request by nithinkrishna**
Rebase before: 38k, now 35k
Increase approval: before 39k, now 38k
Decrease approval: before 32k, now 31k

### CODE REVIEW

**Comment by nithinkrishna** [comment]
## Before change
<pre>
·--------------------------------------------------|----------------------------|-------------|----------------------------·
|               Solc version: 0.4.24               ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 9500000 gas  │
···················································|····························|·············|·····························
|  Methods                                                                                                                 │
·····················|·····························|·············|··············|·············|··············|··············
|  Contract          ·  Method                     ·  Min        ·  Max         ·  Avg        ·  # calls     ·  usd (avg)  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  addTransaction             ·      73038  ·      133597  ·     111666  ·           6  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  rebase                     ·      27556  ·      128221  ·      83603  ·          78  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  removeTransaction          ·      29864  ·       53581  ·      39487  ·           3  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  setTransactionEnabled      ·      25942  ·       30130  ·      28036  ·           2  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  transferOwnership          ·          -  ·           -  ·      31103  ·           1  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  approve                    ·      16680  ·       46752  ·      34728  ·          27  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  decreaseAllowance          ·      29847  ·       34175  ·      32011  ·           8  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  increaseAllowance          ·      33910  ·       48934  ·      39922  ·          10  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  rebase                     ·      28553  ·       42256  ·      38514  ·          13  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  setMonetaryPolicy          ·          -  ·           -  ·      47479  ·           8  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  transfer                   ·      40278  ·       55254  ·      53888  ·          11  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  transferFrom               ·          -  ·           -  ·      46732  ·           2  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setCpiOracle               ·      31312  ·       46312  ·      38812  ·           4  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setDeviationThreshold      ·      15235  ·       30295  ·      18247  ·           5  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setMarketOracle            ·      31532  ·       46532  ·      39032  ·           4  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setOrchestrator            ·      31642  ·       46630  ·      39136  ·           4  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setRebaseLag               ·          -  ·           -  ·      30210  ·           2  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setRebaseTimingParameters  ·      25558  ·       40594  ·      29612  ·          14  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Deployments                                     ·                                          ·  % of limit  ·             │
···················································|·············|··············|·············|··············|··············
|  Orchestrator                                    ·          -  ·           -  ·    1172493  ·      12.3 %  ·          -  │
···················································|·············|··············|·············|··············|··············
|  UFragments                                      ·          -  ·           -  ·    2017035  ·      21.2 %  ·          -  │
···················································|·············|··············|·············|··············|··············
|  UFragmentsPolicy                                ·          -  ·           -  ·    1477380  ·      15.6 %  ·          -  │
·--------------------------------------------------|-------------|--------------|-------------|--------------|-------------·
</pre>

***

## After change

<pre>
·--------------------------------------------------|----------------------------|-------------|----------------------------·
|               Solc version: 0.4.24               ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 9500000 gas  │
···················································|····························|·············|·····························
|  Methods                                                                                                                 │
·····················|·····························|·············|··············|·············|··············|··············
|  Contract          ·  Method                     ·  Min        ·  Max         ·  Avg        ·  # calls     ·  usd (avg)  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  addTransaction             ·      73038  ·      133597  ·     111666  ·           6  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  rebase                     ·      27556  ·      128221  ·      83603  ·          78  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  removeTransaction          ·      29864  ·       53581  ·      39487  ·           3  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  setTransactionEnabled      ·      25942  ·       30130  ·      28036  ·           2  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Orchestrator      ·  transferOwnership          ·          -  ·           -  ·      31103  ·           1  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  approve                    ·      16680  ·       46752  ·      34728  ·          27  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  decreaseAllowance          ·      28881  ·       33209  ·      31045  ·           8  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  increaseAllowance          ·      32957  ·       47981  ·      38969  ·          10  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  rebase                     ·      27771  ·       38784  ·      35460  ·          13  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  setMonetaryPolicy          ·          -  ·           -  ·      47479  ·           8  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  transfer                   ·      40278  ·       55254  ·      53888  ·          11  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragments        ·  transferFrom               ·          -  ·           -  ·      46732  ·           2  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setCpiOracle               ·      31312  ·       46312  ·      38812  ·           4  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setDeviationThreshold      ·      15235  ·       30295  ·      18247  ·           5  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setMarketOracle            ·      31532  ·       46532  ·      39032  ·           4  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setOrchestrator            ·      31642  ·       46630  ·      39136  ·           4  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setRebaseLag               ·          -  ·           -  ·      30210  ·           2  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  UFragmentsPolicy  ·  setRebaseTimingParameters  ·      25558  ·       40594  ·      29612  ·          14  ·          -  │
·····················|·····························|·············|··············|·············|··············|··············
|  Deployments                                     ·                                          ·  % of limit  ·             │
···················································|·············|··············|·············|··············|··············
|  Orchestrator                                    ·          -  ·           -  ·    1172493  ·      12.3 %  ·          -  │
···················································|·············|··············|·············|··············|··············
|  UFragments                                      ·          -  ·           -  ·    1937577  ·      20.4 %  ·          -  │
···················································|·············|··············|·············|··············|··············
|  UFragmentsPolicy                                ·          -  ·           -  ·    1477380  ·      15.6 %  ·          -  │
·--------------------------------------------------|-------------|--------------|-------------|--------------|-------------·

</pre>

**Review by thegostep — APPROVED** [review]
love it

File: contracts/UFragments.sol [hunk 1]
```diff
@@ -122,8 +125,8 @@ contract UFragments is ERC20Detailed, Ownable {
         // ever increased, it must be re-included.
         // _totalSupply = TOTAL_GONS.div(_gonsPerFragment)
```
Comment (brandoniles) [inline contracts/UFragments.sol:126]: Let's update this commented line to assign `newTotalSupply` and move the `_totalSupply = newTotalSupply;` below it?

**Review by brandoniles — COMMENTED** [review]
Wow, nice. Just the one line move.

File: contracts/UFragments.sol (hunk 1 above)
Comment (nithinkrishna) [reply contracts/UFragments.sol]: done 👍

File: contracts/UFragments.sol [hunk 2]
```diff
@@ -94,36 +94,40 @@ contract UFragments is ERC20Detailed, Ownable {
...
+        // We recalculate a newTotalSupply to be in agreement with the _gonsPerFragment
         // conversion rate.
         // This means our applied supplyDelta can deviate from the requested supplyDelta,
         // but this deviation is guaranteed to be < (_totalSupply^2)/(TOTAL_GONS - _totalSupply).
         //
         // In the case of _totalSupply <= MAX_UINT128 (our current supply cap), this
         // deviation is guaranteed to be < 1, so we can omit this step. If the supply cap is
         // ever increased, it must be re-included.
-        // _totalSupply = TOTAL_GONS.div(_gonsPerFragment)
+        // _totalSupply = newTotalSupply
```
Comment (brandoniles) [inline contracts/UFragments.sol:125]: I was thinking
```suggestion
        // newTotalSupply = TOTAL_GONS.div(_gonsPerFragment)
```

Comment (nithinkrishna) [reply contracts/UFragments.sol]: done 👍

**Review by brandoniles — APPROVED** [review]
Thanks!

**Pull request merged**